CLIP+VQGAN/generate.py

```diff
@@ -0,0 +1,650 @@
+        args.image_prompts = [image.strip() for image in args.image_prompts]
+
+    if args.make_video and args.make_zoom_video:
+        print("Warning: Make video and make zoom video are mutually exclusive.")
+        args.make_video = False
+        
+    # Make video steps directory
+    if args.make_video or args.make_zoom_video:
+        if not os.path.exists('steps'):
+            os.mkdir('steps')
+
+    # Fallback to CPU if CUDA is not found and make sure GPU video rendering is also disabled
+    # NB. May not work for AMD cards?
+    if not args.cuda_device == 'cpu' and not torch.cuda.is_available():
+        args.cuda_device = 'cpu'
+        args.video_fps = 0
+        print("Warning: No GPU found! Using the CPU instead. The iterations will be slow.")
+        print("Perhaps CUDA/ROCm or the right pytorch version is not properly installed?")
+    
+
+    # Do it
+    device = torch.device(args.cuda_device)
+    model = load_vqgan_model(args.vqgan_config, args.vqgan_checkpoint).to(device)
+    jit = True if float(torch.__version__[:3]) < 1.8 else False
+    perceptor = clip.load(args.clip_model, jit=jit)[0].eval().requires_grad_(False).to(device)
+
+
+    cut_size = perceptor.visual.input_resolution
+
+    f = 2**(model.decoder.num_resolutions - 1)
+    # Cutout class options:
+    # 'latest','original','updated' or 'updatedpooling'
+    if args.cut_method == 'latest':
+        make_cutouts = MakeCutouts(cut_size, args.cutn, args, cut_pow=args.cut_pow)
+    elif args.cut_method == 'original':
+        make_cutouts = MakeCutoutsOrig(cut_size, args.cutn, args, cut_pow=args.cut_pow)
+    elif args.cut_method == 'updated':
+        make_cutouts = MakeCutoutsUpdate(cut_size, args.cutn, args, cut_pow=args.cut_pow)
+    else:
+        make_cutouts = MakeCutoutsPoolingUpdate(cut_size, args.cutn, args, cut_pow=args.cut_pow)    
+
+
+    toksX, toksY = args.size[0] // f, args.size[1] // f
+    sideX, sideY = toksX * f, toksY * f
+
+    # Gumbel or not?
+    if gumbel:
+        e_dim = 256
+        n_toks = model.quantize.n_embed
+        z_min = model.quantize.embed.weight.min(dim=0).values[None, :, None, None]
+        z_max = model.quantize.embed.weight.max(dim=0).values[None, :, None, None]
+    else:
+        e_dim = model.quantize.e_dim
+        n_toks = model.quantize.n_e
+        z_min = model.quantize.embedding.weight.min(dim=0).values[None, :, None, None]
+        z_max = model.quantize.embedding.weight.max(dim=0).values[None, :, None, None]
+
+
+    # Image initialisation
+    if args.init_image:
+        if 'http' in args.init_image:
+            img = Image.open(urlopen(args.init_image))
+        else:
+            img = Image.open(args.init_image)
+            pil_image = img.convert('RGB')
+            pil_image = pil_image.resize((sideX, sideY), Image.LANCZOS)
+            pil_tensor = TF.to_tensor(pil_image)
+            z, *_ = model.encode(pil_tensor.to(device).unsqueeze(0) * 2 - 1)
```

Comment:
These four lines are repeated in all if/elif statements, consider minimizing code duplication